app/doubleprogressdialog.h

```diff
@@ -0,0 +1,56 @@
+#ifndef DOUBLEPROGRESSDIALOG_H
+#define DOUBLEPROGRESSDIALOG_H
+
+#include <QDialog>
+#include <QProgressBar>
+
+namespace Ui {
+class DoubleProgressDialog;
+}
+
+class DoubleProgressDialog : public QDialog
+{
+    Q_OBJECT
+
+public:
+    explicit DoubleProgressDialog(QWidget *parent = 0);
+    ~DoubleProgressDialog();
+
+    QString getStatus();
+    void setStatus(QString msg);
+
+    class ProgressBarControl {
```

Comment:
Any reason to make this inner-class? 

Most of the functions of `ProgressBarControl` can be done with a plain `QProgressBar`.
setMin() => QProgressBar::setMinimum()
setMax() => QProgressBar::setMaximum() 
setRange() => QProgressBar::setRange() 
setPresision() => why need this if there is already a range?

update progress: `float_progress_value * (max - min)`

Reply:
This was to enable float values while obscuring the implementation details. I also wanted the `progressDialog.major` and `progressDialog.minor` format without modifying the UI directly from other classes (imo, the UI implementation should be separate from the widget's api, if you decide to change how the user interface is displayed, you should not have to change code outside of widget's class). Putting stuff like `float_progress_value * (max - min)` makes it less readable in my opinion. The way I have it, you just give it the percentage complete by default. `setPrecision` sets the number of decimal places that the progress bar shows. If the range is 0-1 and the precision is 2, then there are 100 steps. All of the corresponding QProgressBar functions that you pointed out are int only, which is not practical when you are thinking in terms of percentages.

Reply:
If I set range(1, 2) and setValue(1.5), things screw up. 

Either use max-min or use precision with a uniform 0~1 progress value. It doesn't make sense to have both at the same time. And I thought `ratio * (end - begin)` is a very common way to do distance calculation?

Reply:
> If I set range(1, 2) and setValue(1.5), things screw up.

What exactly screws up? That should still work. Now that I think about it, there's an issue when you change the precision. Regardless of that, doing what you describe should put the progress bar exactly half way as one would expect.

I'll think about this a bit more and continue the discussion tomorrow.

Reply:
Ah, my bad, I didn't notice that you do `convertUnits()` in setMin()/setMax() as well. Apologise.

Reply:
Finally understood what you did. It's mapping an arbitrary range of floating numbers to a range of integers for the progress bar. It's more delicate than I thought.

Reply:
That may be the first time my code has been described as "delicate". Well thank you, glad you like it 😁